### ISSUE TITLE
ca: Vault provider creates a new intermediate on every leader rotation

### ISSUE THREAD

**Issue by dnephin** [hashicorp/consul#11811]
Currently we call [Provider.GenerateIntermediate](https://github.com/hashicorp/consul/blob/v1.10.4/agent/consul/leader_connect_ca.go#L429) in the primary DC every time a new leader is elected (either on startup or leader election). This is not a problem for the Consul or AWS PCA providers because those don't use intermediate certs in the primary.

The Vault provider however, [creates a new intermediate certificate every time](https://github.com/hashicorp/consul/blob/v1.10.4/agent/connect/ca/provider_vault.go#L379-L408).  In a scenario where there are more frequent leader elections, this could lead to a very large list of `IntermediateCerts` being stored and sent to clients unnecessarily.

In the secondary we check `ActiveIntermediate` which prevents the creation of these extra intermediates.

We should do something similar in the primary DC. Instead of generating a new one each time, we should see if one already exists and use it.

I have not written a test to reproduce this behaviour.

**Comment by dnephin** [comment]
Seems worth mentioning that this issue will cause the "secondary CA roots watch" routine to run in all secondary DCs on every leader rotation in the primary DC.  The logic seems correct right now in a secondary and it shouldn't also generate a new intermediate in the secondary as a result. But it's possible this was wrong in some versions and would have also triggered a new intermediate in secondary DCs.

**Comment by lstoll-sfdc** [comment]
I don't think this is completely accurate - while a new intermediate is generated on leader election, the list does not grow. Immediately [before the intermediate is generated](https://github.com/hashicorp/consul/blob/7cb2af17c6757b4084465518773c0c5c8762ef6f/agent/consul/leader_connect_ca.go#L487-L496) the `newCARoot` function is called, which [constructs a fresh copy of the stored state](https://github.com/hashicorp/consul/blob/7cb2af17c6757b4084465518773c0c5c8762ef6f/agent/consul/leader_connect_ca.go#L267-L279). The new intermediate is then [appended on to](https://github.com/hashicorp/consul/blob/7cb2af17c6757b4084465518773c0c5c8762ef6f/agent/consul/leader_connect_ca.go#L534-L539) this empty, freshly initialised root and persisted.

This results in only the active intermediate being passed around. While it could probably be generated less, in practice I don't think it has much impact.

**Comment by lstoll** [comment]
I don't think this is completely accurate - while a new intermediate is generated on leader election, the list does not grow. Immediately [before the intermediate is generated](https://github.com/hashicorp/consul/blob/7cb2af17c6757b4084465518773c0c5c8762ef6f/agent/consul/leader_connect_ca.go#L487-L496) the `newCARoot` function is called, which [constructs a fresh copy of the stored state](https://github.com/hashicorp/consul/blob/7cb2af17c6757b4084465518773c0c5c8762ef6f/agent/consul/leader_connect_ca.go#L267-L279). The new intermediate is then [appended on to](https://github.com/hashicorp/consul/blob/7cb2af17c6757b4084465518773c0c5c8762ef6f/agent/consul/leader_connect_ca.go#L534-L539) this empty, freshly initialised root and persisted.

This results in only the active intermediate being passed around. While it could probably be generated less, in practice I don't think it has much impact.

**Comment by eikenb** [comment]
Just to add a note on how I've tried to reproduce.

I was able to narrow all the CA related setup code under [primaryInitialize](https://github.com/hashicorp/consul/blob/197ed67ea75390b9ef712491fc1dd454bb7d73d6/agent/consul/leader_connect_ca.go#L471) and even more specifically I'm pretty sure that [GenerateIntermediate](https://github.com/hashicorp/consul/blob/197ed67ea75390b9ef712491fc1dd454bb7d73d6/agent/connect/ca/provider_vault.go#L571) contains all the relevant code for the intermediate CA generation.

I asked the Vault team to be sure how I was checking the correct thing in the tests. They specifically pointed me to [listing issuers](https://developer.hashicorp.com/vault/api-docs/secret/pki#list-issuers) as the way to check. This value never grew no matter how many times I ran GenerateIntermidiate or primaryInitialize.

**Comment by eikenb** [comment]
I was playing around with generating mermaid charts from tracebacks, and while it wasn't super useful, I'll include the call stack diagram I generated while figuring out the call path just to get some more use out of it...

```mermaid
flowchart TB
	7[ca.VaultProvider.GenerateIntermediate]
	6[consul.CAManager.primaryInitialize]
	5[consul.CAManager.Initialize]
	4[consul.CAManager.Start]
	3[consul.Server.startConnectLeader]
	2[consul.Server.establishLeadership]
	1[consul.Server.leaderLoop]
	0[consul.Server.monitorLeadership.func1]

	0 --> 1 --> 2 --> 3 --> 4 --> 5 --> 6 --> 7
```

**Comment by dnephin** [comment]
Thanks for looking into it @eikenb!

**Comment by jkirschner-hashicorp** [comment]
Re-opening this issue in light of https://github.com/hashicorp/vault/issues/22980, which suggests that the number of non-expired issuers in the Vault PKI secrets engine is growing. That's still a problem, even if the list of intermediate certs being stored/passed around in Consul isn't growing.
